pkg/controller/tiers/tiers_controller.go

```diff
@@ -0,0 +1,158 @@
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tiers
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/tigera/operator/pkg/common"
+	"github.com/tigera/operator/pkg/render"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/networkpolicy"
+	"github.com/tigera/operator/pkg/render/tiers"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/controller/options"
+	"github.com/tigera/operator/pkg/controller/status"
+	"github.com/tigera/operator/pkg/controller/utils"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// The Tiers controller reconciles system tiers and network policies using the v3 API. v3 resources require the API server to be available.
+// System tiers are reconciled here instead of the installation controller since the API server is not available until after installation completes.
+// Similarly, policies for installation and API server controllers are reconciled here to ensure the API server is available when they are reconciled.
+
+var log = logf.Log.WithName("controller_tiers")
+
+// Add creates a new Tiers Controller and adds it to the Manager.
+// The Manager will set fields on the Controller and Start it when the Manager is Started.
+func Add(mgr manager.Manager, opts options.AddOptions) error {
+	if !opts.EnterpriseCRDExists {
+		// No need to start this controller.
+		return nil
+	}
+
+	tierWatchReady := &utils.ReadyFlag{}
+	policyWatchesReady := &utils.ReadyFlag{}
+
+	reconciler := newReconciler(mgr, opts, tierWatchReady, policyWatchesReady)
+
+	c, err := controller.New("tiers-controller", mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		log.Error(err, "Failed to establish a connection to k8s")
+		return err
+	}
+
+	go utils.WaitToAddTierWatch(networkpolicy.TigeraComponentTierName, c, k8sClient, log, tierWatchReady)
+
+	policyNames := []types.NamespacedName{
+		{Name: tiers.APIServerPolicyName, Namespace: rmeta.APIServerNamespace(operatorv1.TigeraSecureEnterprise)},
```

Comment:
I didn't consider that option but I like it, I think it strikes a better balance than the current implementation. If I understand correctly, it would look something like this:
- Impacted* controllers will check if the tier is available**, and pass this result to their render component(s) as a boolean flag
- The render components will only render policy (i.e. return policy when Objects() is called) if the flag is set to true

Now evaluating how this plays out during an installation:
1. Impacted controllers will not reconcile any NetworkPolicy, and eventually become Available
2. The Tiers controller will create the allow-tigera tier once the APIServer and LicenseKey are available
3. Impacted controllers will need to be 'kicked' via watch to reconcile again, and at this time, their network policies will be included in the objects to reconcile since the tier is available.

My main uncertainty is how the impacted controllers are kicked? They can't establish watches on the Tier, since that would require the APIServer/LicenseKey to be available. What would you suggest?

\* Impacted controller being a controller that should reconcile NetworkPolicy, but blocks the creation of Tier by being a requirement for the APIServer or LicenseKey to be available
\*\* We would need to treat failure to query the tier availability as 'no'

Reply:
> My main uncertainty is how the impacted controllers are kicked? They can't establish watches on the Tier, since that would require the APIServer/LicenseKey to be available. What would you suggest?

Couldn't they use a deferred watch in the same way as other controllers? The only diference would be that instead of returning if the watch isn't ready, the controllers would just not install the policy if the watch wasn't ready.

Reply:
Of course🤦. I think the implementation will be much better off this way, so I'd like to have this refactor as part of the current PR. 

Let me know if/when you are done with your initial review. I'll perform this refactor along with the other requested changes